routing/router.go

```diff
@@ -1503,7 +1514,7 @@ func (r *ChannelRouter) SendPayment(payment *LightningPayment) ([32]byte, *Route
 		// of the channel graph and our past HTLC routing
 		// successes/failures.
 		route, err := paySession.RequestRoute(
-			payment, uint32(currentHeight), finalCLTVDelta,
+			g, payment, uint32(currentHeight), finalCLTVDelta,
```

Comment:
1. Note that in current state, findPath does at least O(n) disk reads, because [this](https://github.com/lightningnetwork/lnd/blob/7ea9e7cf4c0523da992e218d022d84f0ec3cc9df/routing/pathfind.go#L458), it can be changed to use a BBST(or hashing) trivially, however that might affect the overall time mildly.
2. The expected disk reads should be at least something like O(N log N/(log log N)) where N = number of nodes, proven by induction(assuming graphs are sparse(edges <= N log N), there is an edge I can add not in dijkstra tree at each step, let all distances be from node 1) of the top of my head. I’m sure there’s a stricter bound somewhere, I’ll try to find one when I can.

However, I think it’s much easier to just add actual benchmarking tests(it’s not hard to simulate random growth, random changes similar to current autopilot model, etc.). If so, I’ll write that and report some results?

Reply:
That's a good point, and it wouldn't surprise me that reading the whole graph into memory in one batch up front is on average faster. Still, it would be nice to have actual metrics before doing the change. Downloading the current testnet graph and using that to generate some testcases (random edge failures etc) would be a good start, at least now that you got the necessary parsing code in place :)

Reply:
@halseth : Sure, I thought current graph might be a bit too specific(so, multiple similar/different, maybe even larger, would be better), but we can always change it back later, so I'll benchmark for the current one for now.

Reply:
@halseth:
On a random 20% failure rate, old version clocks at 8.94ms, while my version at 8.44ms(insignificant difference). On higher failure rate, my version is slow, Although I am not sure why with such a failure rate the payment is still processed so quickly, maybe I'm doing something wrong(just the benchmark code at https://github.com/nalinbhardwaj/lnd/tree/test). If the stats are correct I'll just shift to using the graphSource version of database graph.

Regarding that, do you know of a quicker way than the one I use to access adjacency list of a node by it's Vertex key? Looking at https://github.com/lightningnetwork/lnd/blob/master/channeldb/graph.go it does seem possible to me, however I'm not familiar with it much, so I'll try to look into it if it can be done.